### PR TITLE
[JENKINS-54250] Disable restarting stages in jenkinsfile optionally

### PR DESCRIPTION
**Pull request by pavlovic-ivan**
* JENKINS issue(s):
    * Fixes https://issues.jenkins.io/browse/JENKINS-54250
* Description:
    * This PR adds a new option that will enable/disable "restart from stage" button in Jenkins UI
* Documentation changes:
    * This is the PR for the change to the docs: https://github.com/jenkins-infra/jenkins.io/pull/5314

I have one question though. I wanted to add tests that will assert groovy file, but i wasn't able to load the resources into the tests. Is there some kind of, let say, framework in place that i have to use, or just use standard Java way of loading resources?

### CODE REVIEW

File: pipeline-model-definition/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/actions/RestartDeclarativePipelineAction.java
Comment (dwnusbaum) [inline pipeline-model-definition/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/actions/RestartDeclarativePipelineAction.java:73]: Probably accidentally left in from debugging: (you could also delete `LOGGER` since it looks like this was the only usage)

```suggestion
```

File: pipeline-model-definition/src/test/java/org/jenkinsci/plugins/pipeline/modeldefinition/DeclarativePipeline2Test.java
Comment (dwnusbaum) [inline pipeline-model-definition/src/test/java/org/jenkinsci/plugins/pipeline/modeldefinition/DeclarativePipeline2Test.java:13]: Any particular reason to create a `RealJenkinsRule`-based test? It doesn't hurt, but it is slower than normal `JenkinsRule` and I am not sure that we care much about the enhanced fidelity in this case.

File: pipeline-model-definition/src/test/resources/options/restartableFromStageOptionSetFalse.groovy
Comment (dwnusbaum) [inline pipeline-model-definition/src/test/resources/options/restartableFromStageOptionSetFalse.groovy:25]: Are these files being used anywhere? Maybe you meant to add a test case to `BasicModelDefTest` (or a new class that extends `AbstractModelDefTest`) that runs these and verifies the behavior?

**Review by dwnusbaum — COMMENTED** [review]
Besides what I noted in https://github.com/jenkinsci/workflow-job-plugin/pull/285#discussion_r931092964, the overall idea seems good. I added some comments.

File: pipeline-model-definition/src/test/resources/options/restartableFromStageOptionSetFalse.groovy
Comment (pavlovic-ivan) [reply pipeline-model-definition/src/test/resources/options/restartableFromStageOptionSetFalse.groovy:25]: Hi @dwnusbaum! I've update my description with question i wanted to ask. You are right. I wanted to add a test, but failed in loading the resources. Is there some kind of, let say, framework in place that i have to use, or just use standard Java way of loading resources?

File: pipeline-model-definition/src/test/java/org/jenkinsci/plugins/pipeline/modeldefinition/DeclarativePipeline2Test.java
Comment (pavlovic-ivan) [reply pipeline-model-definition/src/test/java/org/jenkinsci/plugins/pipeline/modeldefinition/DeclarativePipeline2Test.java:13]: No particular reason. I wasn't aware of the `JenkinsRule`. I am going to use that instead 🙏

File: pipeline-model-definition/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/actions/RestartDeclarativePipelineAction.java
Comment (pavlovic-ivan) [reply pipeline-model-definition/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/actions/RestartDeclarativePipelineAction.java:73]: Correct. I will remove it!

File: pipeline-model-definition/src/test/resources/options/restartableFromStageOptionSetFalse.groovy
Comment (dwnusbaum) [reply pipeline-model-definition/src/test/resources/options/restartableFromStageOptionSetFalse.groovy:25]: Take a look at tests in `BasicModelDefTest` (e.g. https://github.com/jenkinsci/pipeline-model-definition-plugin/blob/2654ca721309dc727b2d7b76c4383724f599606d/pipeline-model-definition/src/test/java/org/jenkinsci/plugins/pipeline/modeldefinition/BasicModelDefTest.java#L558) and `AbstractModelDefTest.expect`. You put your test resources in a specific folder (I do not recall offhand, so compare against other tests, and then call `expected('your_resource').go()` to create a `WorkflowJob` that uses that script and runs it, returning the `WorkflowRun`, and then you can assert things on the `WorkflowRun` as in your existing test.

Comment (pavlovic-ivan) [reply pipeline-model-definition/src/test/resources/options/restartableFromStageOptionSetFalse.groovy:25]: Thanks @dwnusbaum. I really appreciate your help. I'll work on this changes this week with all that you've mentioned in mind.

File: pipeline-model-definition/src/test/java/org/jenkinsci/plugins/pipeline/modeldefinition/DeclarativePipeline2Test.java
Comment (pavlovic-ivan) [reply pipeline-model-definition/src/test/java/org/jenkinsci/plugins/pipeline/modeldefinition/DeclarativePipeline2Test.java:13]: I'll remove the test. Now it looks unnecessary, as other tests should cover if all is good with provided option

**Comment by pavlovic-ivan** [comment]
Hi @dwnusbaum. I am following up on the discussion and suggestions made in the https://github.com/jenkinsci/workflow-job-plugin/pull/285#discussion_r931092964. If i define a class that extends `InvisibleAction`, i will no longer be able to write 

```
options {
    disableRestartFromStage()
}
```

as this is now an actions and not an option any more. Or am i missing something?

**Comment by dwnusbaum** [comment]
Keep your `DisableRestartFromStage extends DeclarativeOption` class that is already here. The new `InvisibleAction` class will just be a replacement for the property you added to `WorkflowJob` in https://github.com/jenkinsci/workflow-job-plugin/pull/285. Maybe call it `DisableRestartFromStageAction`. Then in `Utils.groovy`, where you currently interact with `WorkflowJob.disableRestartFromStage`, you would instead use `WorkflowJob.getAction(DisableRestartFromStageAction.class)`.

**Comment by pavlovic-ivan** [comment]
Ah, got it. So the option will still be recognized, in the `Utils`, i will be just adding or removing the action, and not set the `WorkflowJob` as you've said, ok, i think i've got it :)

**Comment by pavlovic-ivan** [comment]
This PR and [this one](https://github.com/jenkins-infra/jenkins.io/pull/5314) are now ready for review. I also want to create a PR for Blueocean plugin, with the same button disabled, however, i can't seem to use what i did when i run `mvn -f blueocean/pom.xml hpi:run`. I continuously get 1.9.2 version of pipeline-model-definition. How do i use my snapshot?

**Comment by pavlovic-ivan** [comment]
Hi @dwnusbaum. Sorry to bother you. Did you had a chance to maybe take a look at my previous comment?

**Comment by dwnusbaum** [comment]
@pavlovic-ivan Generally speaking, Blue Ocean is not being worked on anymore, see the note on https://www.jenkins.io/doc/book/blueocean/. That said, you can install your snapshot in the advanced tab of the plugin manager once Jenkins is running if you want to test things manually.

I will try to review this PR when I get some time, but it might take a few weeks, sorry! Thanks for working on it!

**Comment by dwnusbaum** [comment]
> Regarding Blue Ocean, does this mean i would be creating a PR for nothing, no one will be looking into it?

Yeah, it would probably be a waste of your time.

That said, since you updated `RestartDeclarativePipelineAction.isRestartEnabled` to consider `DisableRestartFromStageAction`, I would assume that Blue Ocean will automatically disable or hide the relevant buttons/UI elements without any changes. I'm not sure though. It would be pretty unfortunate if Blue Ocean uses its own logic for deciding whether restarting from stages is allowed.

**Comment by pavlovic-ivan** [comment]
I really appreciate your input. Thanks a lot for that!

**Comment by pavlovic-ivan** [comment]
Judging **_only_** by the code itself, i think Blue Ocean plugin is not considering if the pipeline is restartable. This method [here](https://github.com/jenkinsci/blueocean-plugin/blob/master/blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/PipelineNodeImpl.java#L181) checks only for the existence of the action, and not if `RestartDeclarativePipelineAction.isRestartEnabled` is true or false.
```java
@Override
public boolean isRestartable() {
    RestartDeclarativePipelineAction restartDeclarativePipelineAction =
        this.run.getAction( RestartDeclarativePipelineAction.class );
    if (restartDeclarativePipelineAction != null) {
        List<String> restartableStages = restartDeclarativePipelineAction.getRestartableStages();
        if (restartableStages != null) {
            return restartableStages.contains(this.getDisplayName())
                && this.getStateObj() == BlueRun.BlueRunState.FINISHED;
        }
    }
    return false;
}
```

**Comment by olamy** [comment]
@pavlovic-ivan feel free to create a PR for BlueOcean I will have a look

**Comment by pavlovic-ivan** [comment]
> @pavlovic-ivan feel free to create a PR for BlueOcean I will have a look

@olamy thank you very much! sorry for the late late reply, i've been busy with other Jenkins PR

File: pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/Utils.groovy
Comment (dwnusbaum) [inline pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/Utils.groovy:631]: nit: I would use a local variable to avoid looking up the action more than once (untested):
 
```suggestion
                DisableRestartFromStageAction disableRestartFromStageAction = j.getAction(DisableRestartFromStageAction.class)
                if(disableRestartFromStageAction == null && declaredDisableRestartFromStageOption != null){
                    j.addAction(new DisableRestartFromStageAction())
                } else if(disableRestartFromStageAction != null && declaredDisableRestartFromStageOption == null) {
                    j.removeAction(disableRestartFromStageAction)
```

File: pipeline-model-definition/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/actions/RestartDeclarativePipelineAction.java
Comment (dwnusbaum) [inline pipeline-model-definition/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/actions/RestartDeclarativePipelineAction.java:113]: nit: I would inline the method and delete it. `isRestartableFromStage` seems too generic of a name in this context and it seems like it would be confused with `isRestartEnabled`.

```suggestion
                run.getParent().getAction(DisableRestartFromStageAction.class) == null;
```

File: pipeline-model-definition/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/options/impl/DisableRestartFromStage.java
Comment (dwnusbaum) [inline pipeline-model-definition/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/options/impl/DisableRestartFromStage.java:20]:
```suggestion
            return "Disable the ability to restart this Pipeline from a specific stage";
```

File: pipeline-model-definition/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/options/impl/DisableRestartFromStage.java
Comment (dwnusbaum) [inline pipeline-model-definition/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/options/impl/DisableRestartFromStage.java:13]: Did you check whether this works with the directive generator (open a Pipeline job in Jenkins, then click "Pipeline Syntax" in the sidebar and then "Declarative Directive Generator")? I cannot remember if Jelly files need to be added for options like this or not. Since this option does not have any configuration, it might be ok as-is.

File: pipeline-model-definition/src/test/resources/options/restartableFromStageDisabled.groovy
Comment (dwnusbaum) [inline pipeline-model-definition/src/test/resources/options/restartableFromStageDisabled.groovy:40]:
```suggestion
```

File: pipeline-model-definition/src/test/resources/options/restartableFromStageEnabled.groovy
Comment (dwnusbaum) [inline pipeline-model-definition/src/test/resources/options/restartableFromStageEnabled.groovy:37]:
```suggestion
```

File: pipeline-model-definition/src/test/resources/options/restartableFromStageDisabled.groovy
Comment (dwnusbaum) [inline pipeline-model-definition/src/test/resources/options/restartableFromStageDisabled.groovy:4]: Probably best to just delete the license headers here. IDK why we include them in existing test resources.

File: pipeline-model-definition/src/test/java/org/jenkinsci/plugins/pipeline/modeldefinition/OptionsTest.java
Comment (dwnusbaum) [inline pipeline-model-definition/src/test/java/org/jenkinsci/plugins/pipeline/modeldefinition/OptionsTest.java:573]: I think it would be interesting to store the result of `expect`, keep your current assertions, but then add a `DisableRestartFromStageAction` to the parent manually and call `go` a second time to show that it gets removed.

**Review by dwnusbaum — COMMENTED** [review]
Looks good! I added some relatively minor comments.

File: pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/Utils.groovy
Comment (pavlovic-ivan) [reply pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/Utils.groovy]: Valid point. I'll do that

File: pipeline-model-definition/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/actions/RestartDeclarativePipelineAction.java
Comment (pavlovic-ivan) [reply pipeline-model-definition/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/actions/RestartDeclarativePipelineAction.java]: Sounds good

File: pipeline-model-definition/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/options/impl/DisableRestartFromStage.java
Comment (pavlovic-ivan) [reply pipeline-model-definition/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/options/impl/DisableRestartFromStage.java:20]: Got it

File: pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/Utils.groovy
Comment (pavlovic-ivan) [reply pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/Utils.groovy]: Please verify the change

File: pipeline-model-definition/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/actions/RestartDeclarativePipelineAction.java
Comment (pavlovic-ivan) [reply pipeline-model-definition/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/actions/RestartDeclarativePipelineAction.java]: Please verify the change

File: pipeline-model-definition/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/options/impl/DisableRestartFromStage.java
Comment (pavlovic-ivan) [reply pipeline-model-definition/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/options/impl/DisableRestartFromStage.java]: Please verify the change

File: pipeline-model-definition/src/test/resources/options/restartableFromStageDisabled.groovy
Comment (pavlovic-ivan) [reply pipeline-model-definition/src/test/resources/options/restartableFromStageDisabled.groovy]: Please verify the change

File: pipeline-model-definition/src/test/resources/options/restartableFromStageEnabled.groovy
Comment (pavlovic-ivan) [reply pipeline-model-definition/src/test/resources/options/restartableFromStageEnabled.groovy]: Please verify the change

File: pipeline-model-definition/src/test/resources/options/restartableFromStageDisabled.groovy
Comment (pavlovic-ivan) [reply pipeline-model-definition/src/test/resources/options/restartableFromStageDisabled.groovy]: Please verify the change

File: pipeline-model-definition/src/test/java/org/jenkinsci/plugins/pipeline/modeldefinition/OptionsTest.java
Comment (pavlovic-ivan) [reply pipeline-model-definition/src/test/java/org/jenkinsci/plugins/pipeline/modeldefinition/OptionsTest.java:573]: Please verify the change

File: pipeline-model-definition/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/options/impl/DisableRestartFromStage.java
Comment (pavlovic-ivan) [reply pipeline-model-definition/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/options/impl/DisableRestartFromStage.java:13]: No i haven't. Let me do that and i'll get back to you

Comment (pavlovic-ivan) [reply pipeline-model-definition/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/options/impl/DisableRestartFromStage.java:13]: Hmmmm. I can't add an option at all. Not even when i checkout master branch, rebuild and rerun Jenkins... Not sure why at this point. Build was done both with java 8 and then java 11

Comment (pavlovic-ivan) [reply pipeline-model-definition/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/options/impl/DisableRestartFromStage.java:13]: Looks like i will have to add Jelly files. Still figuring out

**Comment by pavlovic-ivan** [comment]
Hi @dwnusbaum . Do you have any updates/comments for this PR maybe?

**Comment by pavlovic-ivan** [comment]
Hi @dwnusbaum. Sorry to bother you but do you have time to take a look at the PR?

**Review by dwnusbaum — APPROVED** [review]

**Comment by kmartens27** [comment]
@pavlovic-ivan & @dwnusbaum it appears that the branch this was originally created on is out of date with the base branch, and therefore cannot be merged.

As this is a plugin repo, a maintainer of the plugin or author of the pull request would need to ensure it is up to date prior to merging.

**Comment by pavlovic-ivan** [comment]
@kmartens27 @dwnusbaum i've synced my fork, and the source branch. All checks are green. Do you feel confident in moving this PR further?

**Comment by pavlovic-ivan** [comment]
Hi @kmartens27 @dwnusbaum. This is just a kind reminder if you can take a look at the PR. It's already approved, can it be merged as well?

**Comment by kmartens27** [comment]
Hi @pavlovic-ivan, thanks for the heads up! Unfortunately, I am not a maintainer of this plugin [listed here](https://plugins.jenkins.io/pipeline-model-definition/), so I am unable to merge this PR.

@dwnusbaum it appears that you are a maintainer of this plugin, would you or another maintainer be able to check in on this?

**Comment by dwnusbaum** [comment]
I have write permission to this repository for historical reasons, so my approval is green, but I do not actively maintain the plugin (I don't think anyone is really actively maintaining it) and cannot commit to following through on things like documentation changes for PRs, investigating possible regressions, etc., so I prefer not to merge PRs. Maybe someone in @jenkinsci/pipeline-model-definition-plugin-developers is interested.

**Comment by pavlovic-ivan** [comment]
Thanks @dwnusbaum . Would you know someone from the @jenkinsci/pipeline-model-definition-plugin-developers i could contact directly? Not sure if anyone will be notified

**Review by olamy — APPROVED** [review]

**Comment by pavlovic-ivan** [comment]
Hi @olamy, thanks for merging this PR! When can i expect this to be released?

**Comment by olamy** [comment]
build is failing so CD doesn't work.

**Comment by olamy** [comment]
@pavlovic-ivan thanks for your patience. I'm not trying to merge other PRs just to get a green build and hoping for a release to happen

**Comment by olamy** [comment]
@pavlovic-ivan release done finally after merging non related PRs to get a green build

**Comment by olamy** [comment]
> Hi @olamy, sorry i was already out when you've posted the first comment. Thanks a lot for all the effort, i will update my other PRs on this subject (Blueocean and jenkins.io). Out of curiosity what was the issue, and how the non related PRs helped to get a successful build? Was there something on my side i could've done?

@pavlovic-ivan no worries. Thanks a lot for your contribution! to release the plugin there is a process in place based on GHA which need to have green build. for some reasons (instability or flaky tests) if any tests fail, the release will not be triggered and it looks releasing manually is complicated almost impossible. So we just have to wait a green build by merging other PRs which will trigger new build of master branch...

**Comment by pavlovic-ivan** [comment]
@olamy thanks a lot for the feedback! i will ask a question in the other PR (for Blueocean). The PR for documentation (jenkins.io) is ready, see [here](https://github.com/jenkins-infra/jenkins.io/pull/5314).

**Pull request merged**
